sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala

```diff
@@ -4077,21 +4077,37 @@ object GpuOverrides extends Logging {
         override val childExprs: Seq[BaseExprMeta[_]] =
           hp.expressions.map(GpuOverrides.wrapExpr(_, this.conf, Some(this)))
 
+        private lazy val hashMode = GpuHashPartitioningBase.hashModeFromCpu(hp, this.conf)
+
         override def tagPartForGpu(): Unit = {
-          val arrayWithStructsHashing = hp.expressions.exists(e =>
-            TrampolineUtil.dataTypeExistsRecursively(e.dataType,
-              {
-                case ArrayType(_: StructType, _) => true
-                case _ => false
-              })
-          )
-          if (arrayWithStructsHashing) {
-            willNotWorkOnGpu("hashing arrays with structs is not supported")
+          this.hashMode match {
+            case HiveMode =>
+              val hh = HiveHash(hp.expressions)
+              val hfMeta = GpuOverrides.wrapExpr(hh, this.conf, None)
+              hfMeta.tagForGpu()
+              if (!hfMeta.canThisBeReplaced) {
+                willNotWorkOnGpu(s"the hash function: ${hh.getClass.getSimpleName}" +
```

Comment:
My only real concern with this patch is in the testing. 

https://github.com/NVIDIA/spark-rapids/blob/branch-25.04/integration_tests/src/main/python/repart_test.py 

is where we test all of our hash partitioning code. It should still work fine if our hash partitioning code matches what the CPU is doing.  My only concern is that the tests are written 100% with murmur3 hashing in mind and with the limitation that we don't support a hash key that is an Array with a Struct under it. So there are no tests for that. Where as the hive hash code does support it, but has a maximum nesting limit instead.  At a minimum can we have a follow on issue to find a proper way to let the integration tests know what hashing mode is being used and add update the integration tests to have proper coverage?

Reply:
Hmm, on top of my mind, there may be two options: 1. introduce a test-only configuration allowing manually configure HiveHash as shuffle partition strategy; 2. cover the integration test with a custom Spark which allows HiveHash as shuffle partition.

Reply:
Concerning option 1, we still can not verify whether partitioning with HiveHash will produce the same result with Spark.

Reply:
Here is the followup issue for tests things .
https://github.com/NVIDIA/spark-rapids/issues/12266